libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -110,70 +134,148 @@ public static bool IsTrustedServiceUrl(string serviceUrl)
+                string token = await this.GetTokenAsync().ConfigureAwait(false);
+                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
             }
-            await base.ProcessHttpRequestAsync(request, cancellationToken);
+            await base.ProcessHttpRequestAsync(request, cancellationToken).ConfigureAwait(false);
         }
 
         public async Task<string> GetTokenAsync(bool forceRefresh = false)
         {
-            if (forceRefresh == false)
+            Task<OAuthResponse> oAuthTokenTask = null;
+            OAuthResponse oAuthToken = null;
+
+            // get tokenTask from cache 
+            lock (tokenTaskCache)
             {
-                // check the global cache for the token. If we have it, and it's valid, we're done. 
-                lock (_cacheSync)
+                // if we are being forced or don't have a token in our cache at all
+                if (forceRefresh || !tokenCache.TryGetValue(TokenCacheKey, out oAuthToken))
                 {
-                    bool foundToken = _cache.TryGetValue(TokenCacheKey, out OAuthResponse oAuthToken);
-                    if (foundToken)
+                    // we will await this task, because we don't have a token and we need it
+                    oAuthTokenTask = _getCurrentTokenTask(forceRefresh: forceRefresh);
+                }
+                else
+                {
+                    // we have an oAuthToken
+                    // check to see if our token is expired 
+                    if (TokenExpired(oAuthToken))
                     {
-                        // we have the token. Is it valid? 
-                        if (oAuthToken.expiration_time > DateTime.UtcNow)
+                        // it is, we should await the current task (someone could have already asked for a new token)
+                        oAuthTokenTask = _getCurrentTokenTask(forceRefresh: false);
+
+                        // if the task is completed and is the expired token, then we need to force a new one 
+                        // (This happens if bot has been 100% idle past the expiration point)
+                        if (oAuthTokenTask.Status == TaskStatus.RanToCompletion && oAuthTokenTask.Result.access_token == oAuthToken.access_token)
                         {
-                            return oAuthToken.access_token;
+                            oAuthTokenTask = _getCurrentTokenTask(forceRefresh: true);
                         }
                     }
+
+                    // always check the autorefresh
+                    CheckAutoRefreshToken();
                 }
             }
 
-            // We need to refresh the token, because:
-            // 1. The user requested it via the forceRefresh parameter
-            // 2. We have it, but it's expired
-            // 3. We don't have it in the cache. 
+            // if we have an oAuthTokenTask then we need to await it
```

Comment:
The newer comments, in the older code, should be brought forward. They're much more descriptive.

Reply:
this code is completely rewritten and this is accurate description of code

---
In reply to: [198676923](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198676923) [](ancestors = 198676923)